app/models/payment_intent.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class PaymentIntent < ApplicationRecord
+  SUCCEEDED_STATES = %w[succeeded].freeze
+
+  belongs_to :invoice, inverse_of: :payment_intents
+
+  validates :invoice, :payment_intent_id, :state, presence: true
+  validates :payment_intent_id, :state, length: { maximum: 255 }
+
+  scope :latest, -> (count = 1) { reorder(created_at: :desc, id: :desc).limit(count) }
+  scope :latest_pending, -> (count = 1) { where.not(state: SUCCEEDED_STATES).latest(count) }
+
+  scope :by_invoice, ->(invoice) { where(invoice: invoice) }
+
+  def succeeded?
+    SUCCEEDED_STATES.include?(state)
+  end
+
+  def update_from_stripe_event(event)
+    return unless event.type == 'payment_intent.succeeded'
+    return if succeeded? || invoice.paid?
+
+    payment_intent_data = event.data.object
+
+    transaction do
```

Comment:
I was thinking, in case of the callback triggered multiple times by Stripe in a short amount of time, there is no lock mechanism, meaning we could have erratic random results:

1. Everything goes fine, invoice is marked as paid, transaction is successful, amount is charged only once
2. Invoice  marked as paid, amount is charged once, transaction failed (double attempt to pay)
3. Invoice marked as failed/unpaid, amount is charged  once, transaction failed
4. etc ...

There is no guarantee that this code will run only once at the same time. Or if it runs multiple times then it is idempotent